### PR TITLE
Refactor post rewriter for addition

### PR DESCRIPTION
**Pull request by nafur**
This makes the post rewriter for addition aware of real algebraic numbers. Also, it tries a bit harder to avoid the `normal_form` utilities if only constants are present.

### CODE REVIEW

File: src/theory/arith/arith_rewriter.cpp [hunk 1]
```diff
@@ -346,27 +347,64 @@ RewriteResponse ArithRewriter::preRewritePlus(TNode t){
 RewriteResponse ArithRewriter::postRewritePlus(TNode t){
   Assert(t.getKind() == kind::PLUS);
 
+  if (t.getNumChildren() == 1)
+  {
+    return RewriteResponse(REWRITE_DONE, t[0]);
+  }
```
Comment (4tXJ7f) [inline src/theory/arith/arith_rewriter.cpp:353]: [It looks like this shouldn't be possible.](https://github.com/cvc5/cvc5/blob/c9daef3594dec48693ede0c13e7fc9804835ef14/src/theory/arith/kinds#L15)

File: src/theory/arith/arith_rewriter.cpp [hunk 2]
```diff
@@ -346,27 +347,64 @@ RewriteResponse ArithRewriter::preRewritePlus(TNode t){
...
   }
 
   if(!monomials.empty()){
     Monomial::sort(monomials);
     Monomial::combineAdjacentMonomials(monomials);
-    polynomials.push_back(Polynomial::mkPolynomial(monomials));
+    polynomials.emplace_back(Polynomial::mkPolynomial(monomials));
   }
+  polynomials.emplace_back(
+      Polynomial::mkPolynomial(Constant::mkConstant(rational)));
```
Comment (4tXJ7f) [inline src/theory/arith/arith_rewriter.cpp:390]: Shouldn't this only be done if `rational` is non-zero? Probably doesn't matter too much, but still.

File: src/theory/arith/arith_rewriter.cpp [hunk 3]
```diff
@@ -346,27 +347,64 @@ RewriteResponse ArithRewriter::preRewritePlus(TNode t){
...
+  auto* nm = NodeManager::currentNM();
+  if (poly.isConstant())
+  {
+    ran += RealAlgebraicNumber(poly.getHead().getConstant().getValue());
+    return RewriteResponse(REWRITE_DONE, nm->mkRealAlgebraicNumber(ran));
+  }
+  return RewriteResponse(
+      REWRITE_DONE,
+      nm->mkNode(
+          Kind::PLUS, nm->mkRealAlgebraicNumber(ran), poly.getNode()));
```
Comment (4tXJ7f) [inline src/theory/arith/arith_rewriter.cpp:407]: Is it safe to do `REWRITE_DONE` here? Couldn't it happen that this gets flattened later on?

**Review by 4tXJ7f — CHANGES_REQUESTED** [review]
Some questions.

File: src/theory/arith/arith_rewriter.cpp (hunk 1 above)
Comment (nafur) [reply src/theory/arith/arith_rewriter.cpp]: Good catch.

File: src/theory/arith/arith_rewriter.cpp (hunk 2 above)
Comment (nafur) [reply src/theory/arith/arith_rewriter.cpp]: true

File: src/theory/arith/arith_rewriter.cpp (hunk 3 above)
Comment (nafur) [reply src/theory/arith/arith_rewriter.cpp]: At this point the poly is not constant and will thus not further rewrite. The only thing that one can further do is merge the constant from poly (if it has one) into the ran. I've added this.

**Review by 4tXJ7f — APPROVED** [review]
LGTM

**Pull request merged**